### PR TITLE
Feat/met 2704 remove mongo jack library

### PR DESCRIPTION
**Pull request by stzanakis**

### CODE REVIEW

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+import eu.europeana.metis.utils.ExternalRequestUtil;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:29]: Unused dependencies?

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  public <T extends MongoTermList<S>, S extends AbstractEdmEntityImpl> MongoTermList<S> findTermListByField(
+      Class<T> mongoTermListType, String fieldName, String fieldValue) {
+    return ExternalRequestUtil.retryableExternalRequestConnectionReset(
+            () -> this.datastore.find(mongoTermListType).filter(fieldName, fieldValue).first());
+  }
+
+
+  private MongoTerm findMongoTermByField(String entityType, String fieldName, String fieldValue) {
+    return ExternalRequestUtil.retryableExternalRequestConnectionReset(
+        () -> this.datastore.find(entityType, MongoTerm.class).filter(fieldName, fieldValue)
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:84]: find is deprecated.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            .first());
+  }
+
+  public List<MongoTerm> getAllMongoTerms(EntityType entityType) {
+    Query<MongoTerm> query = this.datastore.createQuery(getTableName(entityType), MongoTerm.class);
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:89]: createQuery is deprecated.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  public String saveTermList(MongoTermList<? extends AbstractEdmEntityImpl> mongoTermList) {
+    Key<MongoTermList<? extends AbstractEdmEntityImpl>> datasetKey = ExternalRequestUtil
+        .retryableExternalRequestConnectionReset(
+            () -> this.datastore.save(mongoTermList));
+    return datasetKey == null ? StringUtils.EMPTY : datasetKey.getId().toString();
+  }
+
+  public void saveMongoTerms(List<MongoTerm> mongoTerms, EntityType entityType) {
+    ExternalRequestUtil.retryableExternalRequestConnectionReset(
+        () -> this.datastore.save(getTableName(entityType), mongoTerms));
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:116]: save is deprecated.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+  public List<MongoTerm> getAllMongoTerms(EntityType entityType) {
+    Query<MongoTerm> query = this.datastore.createQuery(getTableName(entityType), MongoTerm.class);
+    return getListOfQuery(query, null);
+  }
+
+  public Date getDateOfLastModifiedEntity(EntityType entityType) {
+    Query<MongoTermList> query = datastore.createQuery(MongoTermList.class);
+    query.filter(ENTITY_TYPE_FIELD, getEntityType(entityType));
+    query.order(Sort.descending(MODIFIED_FIELD));
+    final MongoTermList mongoTermList = ExternalRequestUtil
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:97]: SonarLint complains about generics here.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    deleteMongoTermList(codeUri);
+    //Remove from specific collection
+    deleteMongoTerm(entityTable, codeUri);
+
+    //Find all TermLists that have owlSameAs equals with codeUri
+    final Query<MongoTermList> termListsSameAsQuery = this.datastore
+        .createQuery(MongoTermList.class).filter(ENTITY_TYPE_FIELD, entityType)
+        .filter(OWL_SAME_AS, codeUri);
+    final List<MongoTermList> allTermListsSameAs = getListOfQuery(termListsSameAsQuery, null);
+    for (MongoTermList mongoTermList : allTermListsSameAs) {
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:171]: SonarLint complains about generics here.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+      //Remove from specific collection
+      deleteMongoTerm(entityTable, sameAsCodeUri);
+    }
+    return extraUrisRemoved;
+  }
+
+  public void deleteMongoTerm(String entityTable, String codeUri) {
+    final MongoTerm mongoTermCoreUri = findMongoTermByField(entityTable, CODE_URI_FIELD, codeUri);
+    ExternalRequestUtil.retryableExternalRequestConnectionReset(
+        () -> this.datastore.delete(entityTable, MongoTerm.class, mongoTermCoreUri.getId()));
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:185]: delete is deprecated.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+      case ORGANIZATION:
+        result = ORGANIZATION_TYPE;
+        break;
+      default:
+        throw new IllegalStateException("Unknown entity: " + entityType);
+    }
+    return result;
+  }
+
+  private <T> List<T> getListOfQuery(Query<T> query, FindOptions findOptions) {
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:242]: findOptions is always null?

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/MongoTermList.java [hunk 9]
```diff
@@ -1,43 +1,37 @@
...
- * 
+ *
  * @param <T> AgentImpl, PlaceImpl, ConceptImpl, TimespanImpl, OrganizationImpl
+ * @author Yorgos.Mamakis@ europeana.eu
  */
+public abstract class MongoTermList<T extends AbstractEdmEntityImpl> {
 
-public abstract class MongoTermList<T extends AbstractEdmEntityImpl> implements DatedObject {
+  @Id
+  @JsonProperty("_id")
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/MongoTermList.java:21]: I'm assuming it's not required, only maybe when serialized for a view..
I will try to remove it

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/MongoTermList.java [hunk 10]
```diff
@@ -91,29 +77,25 @@ public void setOwlSameAs(String[] owlSameAs) {
...
-  @Override
   public void setCreated(Date created) {
     this.created = created;
   }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends AbstractEdmEntityImpl, S extends T> MongoTermList<T> cast(
+      MongoTermList<S> source) {
+    return (MongoTermList<T>) source;
+  }
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/MongoTermList.java:100]: Correct, not needed anymore

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/RedisProvider.java [hunk 11]
```diff
@@ -84,8 +84,8 @@ private void initPool(String host, int port, String password) {
       }
 
       //Check if connection works
-      try {
-        pool.getResource();
+      try (Jedis jedis = pool.getResource()) {
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/RedisProvider.java:87]: Hahaha that's a good one.. I wonder if there is any other way..

File: metis-enrichment/metis-enrichment-proxy-rest/src/main/java/eu/europeana/enrichment/cache/proxy/config/Application.java [hunk 12]
```diff
@@ -62,8 +67,15 @@ RedisProvider getRedisProvider() {
...
-    return new EnrichmentEntityDao(enrichmentMongo, enrichmentPort);
+  EntityDao getEntityDao() {
+    final MongoProperties<IllegalArgumentException> mongoProperties = new MongoProperties<>(
+        IllegalArgumentException::new);
+    mongoProperties
+        .setAllProperties(new String[]{enrichmentMongoHost}, new int[]{enrichmentMongoPort}, null,
+            null, null, false, null);
+
+    final MongoClient mongoClient = new MongoClientProvider<>(mongoProperties).createMongoClient();
+    return new EntityDao(mongoClient, enrichmentMongoDatabase);
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-proxy-rest/src/main/java/eu/europeana/enrichment/cache/proxy/config/Application.java:78]: I don't think this is required..
We are setting the properties and the dao on each case, more abstraction seems to me an overkill in this case

File: metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/config/Application.java [hunk 13]
```diff
@@ -102,24 +104,22 @@ Enricher enricher() {
...
   @Bean(name = "redisInternalEnricher")
   RedisInternalEnricher getRedisInternalEnricher() {
-    return new RedisInternalEnricher(getEntityDao(), getRedisProvider());
+    final MongoProperties<IllegalArgumentException> mongoProperties = new MongoProperties<>(
+        IllegalArgumentException::new);
+    mongoProperties
+        .setAllProperties(new String[]{enrichmentMongoHost}, new int[]{enrichmentMongoPort}, null,
+            null, null, false, null);
+
+    final MongoClient mongoClient = new MongoClientProvider<>(mongoProperties).createMongoClient();
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/config/Application.java:120]: Same here. I think further abstraction defeats the purpose of those classes

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/EntityService.java [hunk 14]
```diff
@@ -1,12 +1,15 @@
 package eu.europeana.enrichment.service;
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/EntityService.java:1]: I believe it's used somewhere externally, relevant code for the entity api

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 3 above)
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:89]: Yes there is no other way at the moment, we'll need to change the db structure to fix this

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 2 above)
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:84]: Yes there is no other way at the moment, we'll need to change the db structure to fix this

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 1 above)
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:29]: Fixed.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 4 above)
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:116]: Yes there is no other way at the moment, we'll need to change the db structure to fix this

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 7 above)
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:185]: Yes there is no other way at the moment, we'll need to change the db structure to fix this

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 8 above)
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:242]: Sure, simplified

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java [hunk 15]
```diff
@@ -272,38 +279,50 @@ private void loadEntities(EntityType entityType, MongoTerm term, Jedis jedis,
...
-        parentEntities.addAll(this.findParents(parents.getParent(), entityClass));
-      }
-    }
 
+    MongoTermList currentMongoTermList = termList;
+    MongoTermList parentMongoTermList = null;
+    do {
+      if (Objects.nonNull(parentMongoTermList)) {
+        currentMongoTermList = parentMongoTermList;
+      }
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java:305]: Not sure what you mean exactly, can you give an example?

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java [hunk 16]
```diff
@@ -272,38 +279,50 @@ private void loadEntities(EntityType entityType, MongoTerm term, Jedis jedis,
...
-    MongoTermList<?> parents = entityDao.findByCode(parent, entityClass);
-    if (parents != null) {
-      parentEntities.add(parents.getCodeUri());
-      if (parents.getParent() != null && !parent.equals(parents.getParent())) {
-        parentEntities.addAll(this.findParents(parents.getParent(), entityClass));
-      }
-    }
 
+    MongoTermList currentMongoTermList = termList;
+    MongoTermList parentMongoTermList = null;
```
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java:301]: Cannot do much about the generics, but I fixed it in one way..
Same goes for other comments

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 2 above)
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:80]: OK, there is no simple alternative then. Should we create a ticket to do this? This is not going to be done automatically with the ticket that eliminates the redis, right?

File: metis-enrichment/metis-enrichment-proxy-rest/src/main/java/eu/europeana/enrichment/cache/proxy/config/Application.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-proxy-rest/src/main/java/eu/europeana/enrichment/cache/proxy/config/Application.java:78]: I suppose that's a fair point.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java]: This is the kind of fix I was expecting. So no worries.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java:305]: I was thinking of an implementation like this, which seems more compact and perhaps elegant:
```
    final Set<String> parentEntities = new HashSet<>();
    MongoTermList<?> currentMongoTermList = termList;
    while (StringUtils.isNotBlank(currentMongoTermList.getParent())) {
      currentMongoTermList = entityDao.findTermListByField(entityInfo.mongoTermListClass, EntityDao.CODE_URI_FIELD, currentMongoTermList.getParent());
      if (currentMongoTermList == null || !parentEntities.add(currentMongoTermList.getCodeUri())){
        break;
      }
    }
    return parentEntities;
```

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java (hunk 2 above)
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/utils/EntityDao.java:80]: Correct. I created a ticket

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/RedisInternalEnricher.java:305]: Seems identical, I will apply and test

**Pull request merged**